packages/firestore/src/api/user_data_writer.ts

```diff
@@ -158,18 +154,70 @@ export class UserDataWriter {
     const databaseId = new DatabaseId(resourcePath.get(1), resourcePath.get(3));
     const key = new DocumentKey(resourcePath.popFirst(5));
 
-    if (!databaseId.isEqual(this.databaseId)) {
+    if (!databaseId.isEqual(expectedDatabaseId)) {
       // TODO(b/64130202): Somehow support foreign references.
       logError(
         `Document ${key} contains a document ` +
           `reference within a different database (` +
           `${databaseId.projectId}/${databaseId.database}) which is not ` +
           `supported. It will be treated as a reference in the current ` +
-          `database (${this.databaseId.projectId}/${this.databaseId.database}) ` +
+          `database (${expectedDatabaseId.projectId}/${expectedDatabaseId.database}) ` +
           `instead.`
       );
     }
+    return key;
+  }
+
+  protected abstract convertReference(name: string): unknown;
+
+  protected abstract convertBytes(bytes: ByteString): unknown;
+}
+
+export class UserDataWriter extends AbstractUserDataWriter {
+  constructor(protected firestore: Firestore) {
+    super();
+  }
+
+  protected convertBytes(bytes: ByteString): Blob {
+    return new Blob(bytes);
+  }
+
+  protected convertReference(name: string): DocumentReference {
+    const key = this.convertDocumentKey(name, this.firestore._databaseId);
+    return DocumentReference.forKey(key, this.firestore, /* converter= */ null);
```

Comment:
This seems like it now unconditionally creates a reference without a converter, where previously the converter was propagated in from the API construct that was used to create the reference/snapshot. Why is this OK?

Reply:
I thought about this problem, but I think this new behavior actually makes sense. We don't know anything about the document that is referenced here, so it seems like we shouldn't assume that the original converter applies here. I think this is an edge case that probably no one noticed, but it would be pretty challenging to continue exposing the old behavior if we deemed this a breaking change.

Reply:
I think it's reasonable to call the old behavior a bug and call this a fix :-).

We should probably call this out in the release notes.